src/mapper.rs

```diff
@@ -0,0 +1,91 @@
+use rustling_ontology_values::dimension::{Dimension,DatetimeKind};
+use rustling_ontology_values::output::OutputKind;
+pub use rustling_ontology_moment::{Interval, Moment, Local, TimeZone};
+
+pub fn map_dimension(dimension: &mut Dimension, output_kind_filter: &[OutputKind]) {
+
+    match dimension {
+
+        Dimension::Datetime(datetime_value) => {
+
+            // If the filter contains Datetime but no datetime subkind, then no subtyping is required
+            if output_kind_filter.contains(&OutputKind::Datetime) &&
+                !output_kind_filter.contains(&OutputKind::Date) &&
+                !output_kind_filter.contains(&OutputKind::DatePeriod) &&
+                !output_kind_filter.contains(&OutputKind::Time) &&
+                !output_kind_filter.contains(&OutputKind::TimePeriod) {}
+
+            // If Datetime IS NOT in the OutputKind filter, then SOME specific subtyping is
+            // required, through the field datetime_kind of the datetime_value.
+
+            // Technically if the client filter is None, then the filter will contain all possible
+            // output kinds, in which case we want to return more specific subkinds rather than
+            // Datetime.
+
+            // So Datetime will be returned only if Datetime has been
+            // explicitly defined as the filter
+
+            // Helper values to later determine the datetime subkind, if some datetime subkind
+            // hasn't already been set by the grammar. If so then the match will work if the filter
+            // allows for it.
+            if !is_valid_datetime_kind(&datetime_value.datetime_kind) {
+
+                // Find the subkind: Figure out the Datetime subtype from the Form, Grain and other
+                // stuff contained in the Dimension::Datetime(datetime_value)
+
+                let mut date_time_grain = (datetime_value.constraint.grain_left().is_date_grain() &&
+                    datetime_value.constraint.grain_right().is_time_grain()) ||
+                    (datetime_value.constraint.grain_right().is_date_grain() &&
+                        datetime_value.constraint.grain_left().is_time_grain());
+
+                let date_grain = !date_time_grain && datetime_value.constraint.grain_min().is_date_grain();
+
+                let time_grain = datetime_value.is_today_date_and_time() ||
+                    !date_time_grain && datetime_value.constraint.grain_min().is_time_grain();
+
+                let mut has_direction = false;
+                if let Some(_bounded_direction) = datetime_value.direction { has_direction = true };
+                let is_period = datetime_value.has_period_form().unwrap_or(false) ||
+                    datetime_value.has_period_grain().unwrap_or(false) ||
```

Comment:
Why do `. has_period_form()` and `. has_period_grain()` return optionals ?

Reply:
BTW, `is_period` could perhaps be a method of `DatetimeValue`.

Reply:
- The optionals were because at some point some values from the Enum had to return None, but it's not the case after later changes, so these functions return `bool` now.
- Move `is_period` as a DatetimeValue method
✅